### PR TITLE
security(saas): close cross-tenant access surface — login + request-time guards + bug-1 fix

### PR DESCRIPTION
**Pull request by alex-budanov**
Closes the two cross-tenant access bugs surfaced after PR-105 landed:

- **Bug 1**: fake subdomains served `TENANT_EXEMPT_PREFIXES` routes (admin / users-me / audit-logs) because the tenant middleware short-circuited on exempt prefixes before the subdomain lookup
- **Bug 2**: a JWT issued for a user in org A was bearer-equivalent on every tenant subdomain. Login at `orgB.kz.bugspotter.io` with org-A credentials succeeded; reusing any JWT against a different tenant subdomain succeeded; only `assertUserHasActiveOrgAccess` ran, and that check verifies user-has-some-org but not user-belongs-to-this-org

## Mitigation: defense-in-depth (Option D from the design discussion)

| Layer | Where | What it does |
|---|---|---|
| **Tenant resolution tightening** | [`saas/middleware/tenant.ts`](packages/backend/src/saas/middleware/tenant.ts) | Subdomain validation now runs FIRST, even for `TENANT_EXEMPT_PREFIXES`. Fake subdomain on `/admin/*` 404s. Real tenant subdomain on exempt routes still serves (regression-safe — a logged-in user clicking "preferences" inside their dashboard works). Also dropped the `config.public` skip — auth routes are public but absolutely need tenant resolution to fire |
| **Login-time tenant-match guard** | new [`saas/middleware/tenant-match.ts:assertUserBelongsToTenant`](packages/backend/src/saas/middleware/tenant-match.ts) | Wired into login / refresh / magic-login. Before issuing tokens, verifies the user has an org membership matching `request.organizationId`. Same 401 shape as wrong-credentials so attackers can't enumerate. Hub-domain login keeps current product behaviour |
| **Request-time tenant-match middleware** | new [`saas/middleware/tenant-match.ts:createTenantMatchMiddleware`](packages/backend/src/saas/middleware/tenant-match.ts) | Runs after auth + tenant on every authenticated request. Mismatch → 403 `TenantMismatch`. Catches stolen / hub-issued / replayed JWTs |

## Tests

`tests/integration/tenant-cross-access.test.ts` (new, 14 tests) runs the suite in `DEPLOYMENT_MODE=saas`:

- Login at user's own subdomain → 200; at hub → 200; at other tenant → 401 (same shape as wrong-password)
- Login at unknown subdomain → 404 `OrganizationNotFound`
- Refresh: cross-tenant cookie replay → 401; same-tenant → 200
- Magic-login: cross-tenant token redemption → 401 tenant-mismatch; same-tenant → 200
- Request-time: cross-tenant JWT use → 403 `TenantMismatch`; same-tenant → 200; hub-domain → 200
- Bug 1: fake subdomain on /users/me/* → 404 `OrganizationNotFound`; real tenant subdomain serves; hub serves

Full integration suite: 295/295 passing locally (was 281; +14 new). No existing test failed.

## Doc

`packages/backend/docs/auth.md` gets a new §6 covering the three-layer model, what's intentionally NOT enforced (hub login keeps current behaviour, api-key requests skip per existing §5 contract, no platform-admin exemption per product policy), and what's available for further hardening (org-bound JWTs as a follow-up).

## Why register isn't instrumented

Register creates a NEW user with no org binding; the auto-accept invitation step then binds them. Adding a register-time check would require a confusing error message ("registered but can't use this subdomain") for an edge case the request-time middleware already self-corrects on first authenticated use.

## Test plan

- [x] Typecheck clean
- [x] 295/295 integration tests passing locally
- [x] Lint clean
- [ ] CI: integration job + linter gate
- [ ] Reviewer: confirm the no-platform-admin-exemption choice for tenant subdomains matches product intent (per discussion: SaaS admins authenticate at hub only, so an admin JWT on a tenant subdomain is itself anomalous and worth blocking)

## Notes

- **Backward-compat impact**: any token issued at the wrong tenant subdomain will start failing. Since (per product) there are no actual customer accounts yet, blast radius is dev/staging
- **Open question**: org-bound JWTs (bake `organizationId` into the JWT payload, verify on every request) — strongest guarantee but requires forced re-login on rollout. Not in this PR; documented as future hardening in auth.md §7

Generated with Claude Code

## Summary by CodeRabbit

* **Bug Fixes**
  * Enforced tenant membership checks across login, token refresh, magic-link redemption, and authenticated requests in SaaS mode; prevented cross-tenant token replay and JWT bearer equivalence across subdomains. Tenant resolution now runs consistently (including formerly public routes) and returns proper 404/403 for missing or inactive tenants.

* **Documentation**
  * Added SaaS tenant-resolution and cross-tenant guard guidance, plus a hardening note about org-bound JWTs.

* **Tests**
  * Added integration tests covering tenant resolution, cross-tenant rejections, exempt-route behavior, platform-admin cases, and regressions.

### CODE REVIEW

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  errorMessage = 'Invalid email or password'
+): Promise<void> {
+  if (!getDeploymentConfig().features.multiTenancy) {
+    return;
+  }
+  if (!organizationId) {
+    return;
+  }
+  const userOrgs = await db.organizations.findByUserId(user.id);
+  if (!userOrgs.some((o) => o.id === organizationId)) {
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts]: Good call — fixed in 0bc19e4. Switched to `db.organizationMembers.findMembership(organizationId, user.id)` which is an indexed point lookup against the `(organization_id, user_id)` unique key (used by the `ON CONFLICT` in `createWithUser`). Behavior identical, query shape much better.

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    const authUser = request.authUser;
+    const organizationId = request.organizationId;
+    if (!authUser || !organizationId) {
+      return;
+    }
+
+    const userOrgs = await db.organizations.findByUserId(authUser.id);
+    if (userOrgs.some((o) => o.id === organizationId)) {
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts]: Same fix in 0bc19e4. Both call sites now use `findMembership` for an indexed point lookup instead of fetching all orgs and scanning in JS — meaningful for the request-time middleware since it runs on every authenticated tenant request.

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+ *   - `organizationId` is null/undefined (hub-domain login — kept
+ *     working per product decision; users without an org context
+ *     are validated via `assertUserHasActiveOrgAccess` separately)
+ *
+ * @throws AppError(401 'InvalidCredentials') if the user has no
+ *         org membership matching `organizationId`. The 401 shape
+ *         is intentional for login-style routes; callers that want
+ *         a more specific error (e.g., magic-login) should catch
+ *         and re-throw with their own status / code.
+ */
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts]: Real inconsistency, but the fix is the opposite direction — fixed in 038f250 by updating the JSDoc, not the code.

Verified: `sendUnauthorizedWithAttempts` (the wrong-password path this function's contract claims to mirror) sends `error: 'Unauthorized'` ([middleware/auth/responses.ts:92](packages/backend/src/api/middleware/auth/responses.ts#L92)). Every other 401 throw in routes/auth.ts also uses `'Unauthorized'`. Switching this function to `'InvalidCredentials'` would have made it the **only** path emitting that code, which is exactly the diff-on-error-code enumeration vector the comment was trying to prevent.

So the runtime is correct as-is; the JSDoc had the wrong claim. Rewrote the JSDoc to match reality and explain why `'Unauthorized'` is the anti-enumeration choice.

File: packages/backend/src/api/server.ts [hunk 4]
```diff
@@ -406,6 +406,16 @@ export async function createServer(options: ServerOptions): Promise<FastifyInsta
...
 
+  // Register tenant-match middleware (closes the cross-tenant access
+  // surface left open by tenant resolution alone — JWTs aren't org-bound,
+  // so a JWT issued for org A would otherwise be bearer-equivalent on
+  // every tenant subdomain). Registered AFTER auth + tenant so both
+  // `request.authUser` and `request.organizationId` are populated when
+  // it runs. See packages/backend/src/saas/middleware/tenant-match.ts
+  // for the trust contract.
+  const { createTenantMatchMiddleware } = await import('../saas/middleware/tenant-match.js');
+  fastify.addHook('onRequest', createTenantMatchMiddleware(db));
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/server.ts]: Good catch — fixed in 038f250. The hook is now wrapped in a `DEPLOYMENT_MODE.SAAS` check, mirroring `registerTenantMiddleware` (server.ts:92). Self-hosted deployments don't pay the per-request hook-invocation cost anymore (the middleware itself early-returned, but the hook was still being called).

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    // 403 (not 401) because the user IS authenticated; they're just
+    // authenticated for the wrong tenant. 401 would prompt clients
+    // to retry auth, which wouldn't fix the mismatch.
+    sendError(
+      reply,
+      403,
+      'TenantMismatch',
+      "Your account does not have access to this organization's workspace.",
+      request.id
+    );
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts:165]: Done in 038f250. Explicit `return` after `sendError` so any future code added below the rejection branch can't accidentally execute after the response is sent.

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+ * guards, presenting an org-A JWT at `orgB.kz.bugspotter.io` reaches
+ * handlers with `request.organizationId = orgB` and `request.authUser`
+ * = the org-A user, and any handler that doesn't independently
+ * cross-check the two would serve org-B's data.
+ *
+ * Two layers of defence (defense-in-depth):
+ *
+ *   1. Login-time (`assertUserBelongsToTenant`) — auth routes
+ *      (login, register, refresh, magic-login) call this before
+ *      issuing tokens. Stops the wrong JWT from being issued in
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts]: Good catch — fixed in 2d9112e. Verified `/api/v1/auth/register` doesn't call `assertUserBelongsToTenant` (only login at routes/auth.ts:293 and refresh at routes/auth.ts:368 do; magic-login has its own tenant check inline). Comments now reflect that, and explicitly note that register is deliberately uninstrumented because the request-time middleware self-corrects on the very next authenticated request and gating registration would block legitimate cross-tenant signup flows.

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+ */
+export function createTenantMatchMiddleware(db: DatabaseClient) {
+  return async function tenantMatchMiddleware(
+    request: FastifyRequest,
+    reply: FastifyReply
+  ): Promise<void> {
+    const config = getDeploymentConfig();
+    if (config.mode !== DEPLOYMENT_MODE.SAAS) {
+      return;
+    }
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts:131]: Keeping the check intentionally — the per-call cost is one config lookup + one comparison (nanoseconds), and removing it weakens defense-in-depth against future registration drift. Concrete scenarios where the runtime guard catches a mistake the conditional registration alone wouldn't:

- Refactoring `registerTenantMatchMiddleware` to be unconditionally registered "with the runtime check inside it" (e.g., to deduplicate the conditional registration helper across multiple SaaS-only middlewares) — the runtime check is what would still hold the line.
- Tests that import and call `tenantMatchMiddleware` directly without going through `createServer` (none exist today, but the export shape allows it).
- A future caller mounting the middleware on a subroute via `fastify.register` after `createServer` returned, bypassing the conditional in server.ts.

Each of those is unlikely but real, and the cost is genuinely negligible. So treating this as belt-and-suspenders rather than removing it. Open to revisiting if a profile actually surfaces this as a hot path.

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+    const authUser = request.authUser;
+    const organizationId = request.organizationId;
+    if (!authUser || !organizationId) {
+      return;
+    }
+
+    const membership = await db.organizationMembers.findMembership(organizationId, authUser.id);
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts:151]: Same conclusion as the earlier coderabbit nitpick on this — deferring the cache, but worth being explicit about why:

**Why deferred, not declined:**
- The current query is now an indexed point lookup against the `(organization_id, user_id)` unique key (fixed in 0bc19e4 from a fetch-all + JS scan). Tens of microseconds at PG-side, no JOIN — meaningful headroom before this is the bottleneck on a SaaS request path.
- A real cache here needs invalidation on every `organization_members` mutation (add / remove / role change). Plus tombstoning so a removed member can't keep accessing a tenant subdomain via stale cache. Building that correctly is a focused PR's worth of work, with a real correctness risk if invalidation is wrong (cached "yes" after revocation = security regression).
- Scale today doesn't justify it. When request rate or DB latency makes this measurable, profiling will tell us where to put the cache (in-process LRU vs. Redis, TTL vs. event-driven invalidation) — design choices that depend on actual traffic shape, not speculation.

So the targeted-lookup fix is the right scope for this PR; caching belongs to a future perf PR with its own design pass on invalidation semantics. Tracking the suggestion in the PR description's follow-ups list.

File: packages/backend/tests/integration/tenant-cross-access.test.ts [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+          authorization: `Bearer ${userAJwt}`,
+        },
+      });
+
+      expect(response.statusCode).toBe(200);
+      // Project ref consumed elsewhere (kept to avoid unused-var lint).
+      void userAOrgAProject;
+    });
+  });
+
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/tests/integration/tenant-cross-access.test.ts:395]: Real gap, fixed in 2d9112e. Two regression tests added under `Request-time tenant-match middleware`:

1. `rejects a platform-admin JWT at a tenant subdomain the admin is not a member of` — request-time path. Creates a user with `security.is_platform_admin = true` (set via raw UPDATE so generateAuthTokens issues a JWT with `isPlatformAdmin: true`), hub-logins (succeeds because `assertUserHasActiveOrgAccess` exempts platform admins), then sends the JWT to orgA — must return 403 TenantMismatch.
2. `rejects platform-admin login at a tenant subdomain (login-time guard, no exemption)` — same divergence at the login-time guard. POST /auth/login at orgA's subdomain with a platform-admin user not in orgA → 401 Unauthorized (matches the wrong-password shape per the anti-enumeration design).

Note one correction to your snippet: `role: 'admin'` is application-level (`'admin' | 'user' | 'viewer'`), not platform-admin — the platform-admin flag lives in `security.is_platform_admin` (db/types.ts:58). `role: 'admin'` alone wouldn't trigger an `isPlatformAdmin` exemption if one were added, so the test would have passed even with the regression. The corrected fixture sets the security column directly, then the regular login flow (generateAuthTokens reads from there) issues a JWT that does carry `isPlatformAdmin: true` — which is what the regression-guard test needs.

16/16 tenant-cross-access tests passing locally with the additions.

File: packages/backend/src/saas/middleware/tenant-match.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts]: Thanks for the explicit acknowledgement and the Learnings entry — that's the right outcome. The wrong-password / tenant-mismatch / refresh-rejection paths all converging on `'Unauthorized'` is load-bearing for the no-enumeration property; happy to have the rule pinned for future reviews.

File: packages/backend/src/saas/middleware/tenant-match.ts [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant-match.ts:151]: Same suggestion you raised earlier in this PR (now-resolved thread on the same line) — keeping the same answer, with the same reasoning:

The current query is an indexed point lookup against the `(organization_id, user_id)` unique key (fixed in 0bc19e4 from a fetch-all + JS scan). At PG-side, that's tens of microseconds, no JOIN — meaningful headroom before this becomes the bottleneck.

A correct cache here needs:
- Invalidation on every `organization_members` mutation (add / remove / role change) — event-driven, not just TTL, because `real-time access revocation` is what makes this a security control rather than a courtesy
- Tombstoning so a revoked member can't keep accessing via stale cache between mutation and invalidation

Building that correctly is a focused PR's worth of work, with a real correctness risk if invalidation is wrong (cached "yes" after revocation = security regression). Scale today doesn't justify that complexity. When request rate or DB latency makes this measurable, profiling will tell us where to put the cache (in-process LRU vs. Redis, TTL vs. event-driven invalidation) — design choices that depend on actual traffic shape, not speculation.

So: targeted-lookup fix lands here; caching belongs to a future perf PR with its own design pass on invalidation semantics.

File: packages/backend/src/saas/middleware/tenant.ts [hunk 11]
```diff
@@ -110,52 +110,74 @@ export function createTenantMiddleware(db: DatabaseClient) {
...
-      );
-      return;
-    }
-
-    if (!ACTIVE_STATUSES.has(organization.subscription_status)) {
-      sendError(reply, 403, 'SubscriptionInactive', 'Your subscription is not active', request.id, {
-        status: organization.subscription_status,
-      });
+    if (subdomain) {
+      const organization = await db.organizations.findBySubdomain(subdomain);
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/saas/middleware/tenant.ts:136]: Deferred for the same defense-in-depth + correctness reasons as the tenant-match cache suggestion in this PR, plus one extra concern that's specific to this code path:

This middleware doesn't just resolve subdomain → org id. It also enforces the subscription gate at [tenant.ts:147](../src/saas/middleware/tenant.ts#L147):

```ts
if (!ACTIVE_STATUSES.has(organization.subscription_status)) {
  sendError(reply, 403, 'SubscriptionInactive', ...);
}
```

Caching the `organizations` row caches `subscription_status` along with it. Concrete failure modes a naive TTL cache would create:

- **Past-due / cancelled / suspended subscription keeps serving for the cache TTL.** Billing wants the revocation to be near-instant — letting an unpaid customer continue to make tenant-subdomain requests for, say, 60s after their plan flips is a billing/access-control regression, not a perf trade-off.
- **Soft-deleted org keeps resolving.** Same problem at higher severity.
- **Subdomain rename or transfer** — same staleness window.

So the same caveat as elsewhere applies (need event-driven invalidation on mutations, not just TTL), but here `subscription_status` makes it stricter: the cache must be invalidated on every subscription state transition, not just `organizations` table writes. That's coordinated-state work spanning billing, the org table, and the cache — a focused design pass rather than a drop-in.

Today's load doesn't justify it; happy to revisit with profile data and an invalidation design when it does. Tracking on the perf-followup list rather than letting it ride here.

File: packages/backend/src/api/server.ts [hunk 12]
```diff
@@ -406,6 +406,23 @@ export async function createServer(options: ServerOptions): Promise<FastifyInsta
...
+  // `request.authUser` and `request.organizationId` are populated when
+  // it runs. SaaS-only — same pattern as `registerTenantMiddleware`,
+  // so self-hosted deployments don't pay the per-request hook cost.
+  // See packages/backend/src/saas/middleware/tenant-match.ts for the
+  // trust contract.
+  {
+    const { getDeploymentConfig, DEPLOYMENT_MODE } = await import('../saas/config.js');
+    if (getDeploymentConfig().mode === DEPLOYMENT_MODE.SAAS) {
+      const { createTenantMatchMiddleware } = await import('../saas/middleware/tenant-match.js');
+      fastify.addHook('onRequest', createTenantMatchMiddleware(db));
```
[the remark this answers is not in the document]
Comment (alex-budanov) [reply packages/backend/src/api/server.ts:422]: Pushing back on this one — the claim doesn't hold for Fastify v5, and the suggested change would actually weaken the wiring. Verified four ways:

**1. Fastify v5 type contract.** [`node_modules/.pnpm/fastify@5.8.5/.../types/request.d.ts:83`](https://github.com/fastify/fastify/blob/main/types/request.d.ts) declares `routeOptions` as `readonly routeOptions: Readonly<RequestRouteOptions<...>>` — non-optional. By design, route resolution happens *before* the matched route's hooks fire (including `onRequest`), so `routeOptions` is populated by the time any hook on a matched route runs.

**2. Existing tenant middleware uses the exact same pattern.** `createTenantMiddleware` ([tenant.ts:124](../src/saas/middleware/tenant.ts#L124), pre-existing) is also registered as `onRequest` and reads `request.routeOptions?.url`. It's been live in production since SaaS mode shipped; if `routeOptions` were undefined in `onRequest`, that middleware would have skipped its checks for every production request and the entire tenant-resolution layer would have been a no-op since day one.

**3. The reject-cross-tenant test specifically depends on `routeOptions.url` being populated.** `tests/integration/tenant-cross-access.test.ts:279` (`rejects an authenticated request to a tenant subdomain that the user does not belong to`) hits `/api/v1/projects` with a JWT for a user not in the target org and asserts `403 TenantMismatch`. The middleware's first early-return is `if (!url) return;` — if `routeOptions.url` were undefined, the test would get 200 (the project list) instead of 403. Just re-ran the test in isolation: passes in 485ms.

**4. `preValidation` would actually be wrong here.** The middleware needs to run *after* `createAuthMiddleware` so `request.authUser` is populated for the membership check ([server.ts:404](../src/api/server.ts#L404)). Auth runs at `onRequest`. Moving tenant-match to `preValidation` would still work for that specific dependency (preValidation is later), but it'd also delay the 403 past Fastify's body-parsing stage — meaning a cross-tenant POST would have its (potentially large, potentially malicious) body parsed before being rejected. The earlier we reject in the lifecycle, the smaller the attack surface and the smaller the wasted-CPU bill on rejected traffic.

So: not changing this, not a regression. Happy to add a one-line dev-time `assert(request.routeOptions, 'expected routeOptions in onRequest')` if you want explicit lock-in against a Fastify-version regression that flips this contract; let me know if that'd be useful.
